vm/ubpf_loader.c

```diff
@@ -149,154 +182,305 @@ ubpf_load_elf(struct ubpf_vm* vm, const void* elf, size_t elf_size, char** errms
         sections[i].size = shdr->sh_size;
     }
 
-    /* Find first text section */
-    int text_shndx = 0;
-    for (i = 0; i < ehdr->e_shnum; i++) {
+    const char* strtab_data = NULL;
+    int strtab_size = 0;
+    for (i = 0; i < section_count; i++) {
         const Elf64_Shdr* shdr = sections[i].shdr;
-        if (shdr->sh_type == SHT_PROGBITS && shdr->sh_flags == (SHF_ALLOC | SHF_EXECINSTR)) {
-            text_shndx = i;
+        if (shdr->sh_type == SHT_STRTAB) {
+            strtab_data = sections[i].data;
+            strtab_size = sections[i].size;
             break;
         }
     }
 
-    if (!text_shndx) {
-        *errmsg = ubpf_error("text section not found");
+    if (!strtab_data) {
+        *errmsg = ubpf_error("could not find the string table in the elf file");
         goto error;
     }
 
-    struct section* text = &sections[text_shndx];
+    Elf64_Sym* symbols = NULL;
+    int symtab_size = 0;
+    for (i = 0; i < section_count; i++) {
+        const Elf64_Shdr* shdr = sections[i].shdr;
+        if (shdr->sh_type == SHT_SYMTAB) {
+            symbols = (Elf64_Sym*)sections[i].data;
+            symtab_size = sections[i].size;
+            break;
+        }
+    }
 
-    /* May need to modify text for relocations, so make a copy */
-    text_copy = malloc(text->size);
-    if (!text_copy) {
-        *errmsg = ubpf_error("failed to allocate memory");
+    if (!symbols) {
+        *errmsg = ubpf_error("could not find the symbol table in the elf file");
         goto error;
     }
-    memcpy(text_copy, text->data, text->size);
 
-    /* Process each relocation section */
-    for (i = 0; i < ehdr->e_shnum; i++) {
-        struct section* rel = &sections[i];
-        if (rel->shdr->sh_type != SHT_REL) {
-            continue;
-        } else if (rel->shdr->sh_info != text_shndx) {
+    uint64_t total_symbols = symtab_size / sizeof(Elf64_Sym);
+    uint linked_program_size = 0;
+
+    /*
+     * Be conservative and assume that each of the symbols represents a function.
+     */
+    relocated_functions = (struct relocated_function**)calloc(total_symbols, sizeof(struct relocated_function*));
+    total_functions = 1;
+    for (uint64_t i = 0; i < total_symbols; i++) {
+        const Elf64_Sym* sym = symbols + i;
+
+        if (ELF64_ST_TYPE(sym->st_info) != STT_FUNC) {
             continue;
         }
 
-        const Elf64_Rel* rs = rel->data;
+        struct relocated_function* rf = (struct relocated_function*)calloc(1, sizeof(struct relocated_function));
```

Comment:
It looks like this is leaked on error?

Reply:
Also, we need to handle calloc failing on out of memory.

Reply:
Good eyes!!

Reply:
Let me know if the latest version looks better!